base/expr.jl

```diff
@@ -442,3 +445,103 @@ macro generated(f)
         error("invalid syntax; @generated must be used with a function definition")
     end
 end
+
+
+"""
+    @atomic var
+    @atomic order ex
+
+Mark `var` or `ex` as being performed atomically, if `ex` is a supported expression.
+
+See [atomics](#man-atomics) in the manual for more details.
+
+```jldoctest
+julia> mutable struct Atomic{T}; @atomic x::T; end
+
+julia> a = Atomic(1)
+Atomic{Int64}(1)
+
+julia> @atomic :sequentially_consistent a.x = 2 # set field x of a, with sequential consistency
+2
+
+julia> @atomic a.x # fetch field x of a, with sequential consistency
+2
+
+julia> @atomic a.x += 1 # increment field x of a, with sequential consistency
+3
+
+julia> y = 4; @atomic a.x, z = y, a.x # swap field x of a with y and put the old value in z
+(4, 3)
+
+julia> z
+3
+
+julia> y = 5; @atomic a.x, z = y, _ # swap field x of a with y and put the old value in z
+(5, 4)
+
+julia> z
+4
+```
+
+The following forms are also planned, but not yet implemented:
```

Comment:
I would just leave this out; no need to count chickens before they hatch.

Reply:
I possibly just need to finish that. The code is mostly present below to detect it, just not the final expansion.

Reply:
Hm, this seems to be syntactically ambiguous too now with setproperty, which needs to be addressed.

And as a TODO: need to also be clear here not to duplicate the load you meant to be atomic, as it won't be fused into the atomic store